README
Name: The vc bundle
Version: v0.6
Authors: Roland Hieber <rohieb+ctan@rohieb.name>,
  Stephan Hennig <stephanhennig@arcor.de>
Last modified: 2018-06-11
License: CC0 1.0 (Public Domain Dedication), see ./LICENSE,
  or <URL:http://creativecommons.org/publicdomain/zero/1.0/>


The vc bundle
-----------------------------------
This is a script based approach to version control for TeX
documents.  It works more reliably than keyword substitution
based approaches, since it tracks all files in a working
copy, not only .tex files.

The vc bundle works with LaTeX and plain TeX.  Currently,
Bazaar, Git, Mercurial and Subversion are supported.


What's special about the vc bundle?
-----------------------------------
* Common version control packages for LaTeX use keyword
  substitution to access revision information.  That
  approach doesn't work reliably in the presence of
  non-.tex files, e.g., graphics or graphic sources that
  are processed by an external tool.
* The author is unaware of any alternative version control
  packages for plain TeX.
* The author is unaware of any alternative LaTeX version
  control packages that support Bazaar or Git.


Getting the software
-----------------------------------
The latest version of the vc bundle can be found at
CTAN:support/vc.

<URL:http://www.ctan.org/get/support/vc.zip>

The source repository can be found on GitHub.

<URL:https://github.com/rohieb/ctan-support-vc>


Installation and Usage
-----------------------------------
Please see the manual in vc-manual.pdf.

Happy TeXing!
Stephan Hennig


CHANGES
v0.6 (2018-06-11)
----------------------
Changes:
* Meta:
  - New maintainer: Roland Hieber <rohieb+ctan@rohieb.name>
  - Move code to GitHub: https://github.com/rohieb/ctan-support-vc
* Add support for Mercurial, thanks to ildar nizamov (@niya3 on GitHub)
* Git: parse the machine-readable output of `git status --porcelain`
  - fixes the issue that local modifications were not detected in recent Git
    versions
* Git: use committer date instead of author date for \VCDate, since author date
  is not guaranteed to be monotonous
* Some code cleanup


v0.5 (2008-04-21)
----------------------
Incompatible changes:
* Don't write sensitive information to vc.tex, by default (see appendix A).
  - Bazaar:
      \BZRBranchNick
  - Subversion:
      \SVNPath, \SVNName, \SVNUrl, \SVNNodeKind, \SVNRepositoryRoot


v0.42 (2008-04-17)
----------------------
Changes:
* Small documentation update.


v0.41 (2008-04-11)
----------------------
Changes:
* Small documentation update.


v0.4 (2008-04-10)
----------------------
Incompatible changes:
* Search for local modifications only at request
  (relevant to macro \VCRevisionMod, see appendix B).

Changes:
* Again improved locale handling.
* Improved compatibility for Subversion within Msys shell.
* Improved documentation:


v0.3 (2008-03-20)
----------------------
New:
* Added support for Bazaar.
* The vc bundle is now Public Domain.

Changes:
* Fixed line endings in Unix scripts (solving several issues).
* Fixed handling of variable LANG in shell scripts.
* Improved documentation:
  - Added some more "Questions and answers".


v0.21 (2008-03-10)
----------------------
* Incompatible change:
  - Removed macros \VCDate, \VCLongDate, \VCDateTeX.
  - Introduced macros \VCDateRAW, \VCDateISO, \VCDateTEX.
* Improved documentation:
  - Added section "Questions and answers".


v0.2 (2008-03-06)
----------------------
* Added documentation.
* First upload to CTAN.


v0.1alpha (2008-03-04)
----------------------
* First public announcement on comp.text.tex.


make-release.sh
#!/bin/sh -e
if [ -z "$1" ]; then
	echo "Usage: $0 <version>"
	exit
fi

version=$1
date=`date +%Y-%m-%d`

sed -i 's/^\(Version:\).*$/\1 v'"$version"'/' README
sed -i 's/^\(Last modified:\).*$/\1 '"$date"'/' README
sed -i 's/^\(\\newcommand\*{\\pkgversion}\).*$/\1{v'"$version"'}/' vc-manual.tex

tmpfile=`mktemp`
(
	echo "v$version ($date)"
	echo "----------------------"
	echo "* TODO: Describe the changes here."
	echo
	echo
	cat CHANGES
) > $tmpfile
mv $tmpfile CHANGES
sensible-editor CHANGES

git add CHANGES README vc-manual.tex
git commit -sm "release v$version"

cd git-unix/
./vc -m
mv vc.tex ..
cd ..

pdflatex vc-manual.tex
pdflatex vc-manual.tex

git add vc-manual.pdf
git commit -sm "update manual for version $version"
git tag -sm "release v$version" v$version

git archive -v -o vc-v$version.zip --prefix=vc/ v$version


bzr-unix/vc-bzr.awk
# This is file 'vc-bzr.awk' from the vc bundle for TeX.
# The original file can be found at CTAN:support/vc.
# This file is Public Domain.
BEGIN {
### Assume clean working copy.
		Clean = "True"
}
/^revision-id:/ { RevisionId = substr($0, 2+match($0, ":")) }
/^date:/ { Date = substr($0, 2+match($0, ":")) }
/^build-date:/ { BuildDate = substr($0, 2+match($0, ":")) }
/^revno:/ { RevNo = substr($0, 2+match($0, ":")) }
/^branch-nick:/ { BranchNick = substr($0, 2+match($0, ":")) }
/^clean:/ { Clean = substr($0, 2+match($0, ":")) }
END {
### Extract relevant information from variables.
		elements = split(RevisionId, elem, "-")
		Author = elem[1]
		for (i=2; i<elements-2; i++) {
				Author = Author "-" elem[i]
		}
    LongDate = substr(Date, 1, 25)
    DateRAW = substr(LongDate, 1, 10)
    DateISO = DateRAW
    DateTEX = DateISO
    gsub("-", "/", DateTEX)
    Time = substr(LongDate, 12, 14)
		if (Clean=="True") {
				modified = 0
		} else {
				modified = 1
		}
### Write file identification to vc.tex.
		print "%%% This file has been generated by the vc bundle for TeX."
		print "%%% Do not edit this file!"
		print "%%%"
### Write Bazaar specific macros.
    print "%%% Define Bazaar specific macros."
    print "\\gdef\\BZRRevisionId{" RevisionId "}%"
    print "\\gdef\\BZRDate{" Date "}%"
    print "\\gdef\\BZRBuildDate{" BuildDate "}%"
    print "\\gdef\\BZRRevNo{" RevNo "}%"
		if (full==1) {
				print "\\gdef\\BZRBranchNick{" BranchNick "}%"
		}
### Write generic version control macros.
    print "%%% Define generic version control macros."
    print "\\gdef\\VCRevision{\\BZRRevNo}%"
    print "\\gdef\\VCAuthor{" Author "}%"
    print "\\gdef\\VCDateRAW{" DateRAW "}%"
    print "\\gdef\\VCDateISO{" DateISO "}%"
    print "\\gdef\\VCDateTEX{" DateTEX "}%"
    print "\\gdef\\VCTime{" Time "}%"
    print "\\gdef\\VCModifiedText{\\textcolor{red}{with local modifications!}}%"
### Is working copy modified?
    print "%%% Is working copy modified?"
    print "\\gdef\\VCModified{" modified "}%"
		if (modified==0) {
				print "\\gdef\\VCRevisionMod{\\VCRevision}%"
		} else {
				print "\\gdef\\VCRevisionMod{\\VCRevision~\\VCModifiedText}%"
		}
}


hg-windows/vc-hg.awk
# This is file 'vc-hg.awk' from the vc bundle for TeX.
# The original file can be found at CTAN:support/vc.
# This file is Public Domain.
BEGIN {
		
### Process output of "hg status".
		if (script=="status") {
				modified = 0
		}

}



### Process output of "hg log".
script=="log" && /^Hash:/ { Hash = substr($0, 2+match($0, ":")) }
script=="log" && /^Abr. Hash:/ { AbrHash = substr($0, 2+match($0, ":")) }
script=="log" && /^Branch:/ { Branch = substr($0, 2+match($0, ":")) }
script=="log" && /^Parent1 Hash:/ { Parent1Hash = substr($0, 2+match($0, ":")) }
script=="log" && /^Parent2 Hash:/ { Parent2Hash = substr($0, 2+match($0, ":")) }
script=="log" && /^Abr. Parent1 Hash:/ { AbrParent1Hash = substr($0, 2+match($0, ":")) }
script=="log" && /^Abr. Parent2 Hash:/ { AbrParent2Hash = substr($0, 2+match($0, ":")) }
script=="log" && /^Author Name:/ { AuthorName = substr($0, 2+match($0, ":")) }
script=="log" && /^Author Email:/ { AuthorEmail = substr($0, 2+match($0, ":")) }
script=="log" && /^Author Date:/ { AuthorDate = substr($0, 2+match($0, ":")) }

### Process output of "hg status".
script=="status" && /^M / { modified = 1 }



END {
		
### Process output of "hg log".
		if (script=="log") {

    ### Standard encoding is UTF-8.
            if (Encoding == "") Encoding = "UTF-8"
    ### Extract relevant information from variables.
            LongDate = substr(AuthorDate, 1, 25)
            DateRAW = substr(LongDate, 1, 10)
            DateISO = DateRAW
            DateTEX = DateISO
            gsub("-", "/", DateTEX)
            Time = substr(LongDate, 12, 14)
    ### Write file identification to vc.tex.
            print "%%% This file has been generated by the vc bundle for TeX."
            print "%%% Do not edit this file!"
            print "%%%"
    ### Write Hg specific macros.
            print "%%% Define HG specific macros."
            print "\\gdef\\HGHash{" Hash "}%"
            print "\\gdef\\HGAbrHash{" AbrHash "}%"
            print "\\gdef\\HGBranch{" Branch "}%"
            print "\\gdef\\HGFirstParentHash{" Parent1Hash "}%"
            print "\\gdef\\HGSecondParentHash{" Parent2Hash "}%"
            print "\\gdef\\HGAbrFirstParentHash{" AbrParent1Hash "}%"
            print "\\gdef\\HGAbrSecondParentHash{" AbrParent2Hash "}%"
            print "\\gdef\\HGAuthorName{" AuthorName "}%"
            print "\\gdef\\HGAuthorEmail{" AuthorEmail "}%"
            print "\\gdef\\HGAuthorDate{" AuthorDate "}%"
    ### Write generic version control macros.
            print "%%% Define generic version control macros."
            print "\\gdef\\VCRevision{\\HGAbrHash}%"
            print "\\gdef\\VCAuthor{\\HGAuthorName}%"
            print "\\gdef\\VCDateRAW{" DateRAW "}%"
            print "\\gdef\\VCDateISO{" DateISO "}%"
            print "\\gdef\\VCDateTEX{" DateTEX "}%"
            print "\\gdef\\VCTime{" Time "}%"
            print "\\gdef\\VCModifiedText{\\textcolor{red}{with local modifications!}}%"
            print "%%% Assume clean working copy."
            print "\\gdef\\VCModified{0}%"
            print "\\gdef\\VCRevisionMod{\\VCRevision}%"
		}
		
### Process output of "hg status".
		if (script=="status") {
				print "%%% Is working copy modified?"
				print "\\gdef\\VCModified{" modified "}%"
				if (modified==0) {
						print "\\gdef\\VCRevisionMod{\\VCRevision}%"
				} else {
						print "\\gdef\\VCRevisionMod{\\VCRevision~\\VCModifiedText}%"
				}
		}
		
}


git-windows/vc-git.awk
# This is file 'vc-git.awk' from the vc bundle for TeX.
# The original file can be found at CTAN:support/vc.
# This file is Public Domain.

BEGIN {
	modified = 0
}

### Process output of "git log"
script == "log" && /^Hash:/            { Hash            = substr($0, 2+match($0, ":")) }
script == "log" && /^AbrHash:/         { AbrHash         = substr($0, 2+match($0, ":")) }
script == "log" && /^ParentHashes:/    { ParentHashes    = substr($0, 2+match($0, ":")) }
script == "log" && /^AbrParentHashes:/ { AbrParentHashes = substr($0, 2+match($0, ":")) }
script == "log" && /^AuthorName:/      { AuthorName      = substr($0, 2+match($0, ":")) }
script == "log" && /^AuthorEmail:/     { AuthorEmail     = substr($0, 2+match($0, ":")) }
script == "log" && /^AuthorDate:/      { AuthorDate      = substr($0, 2+match($0, ":")) }
script == "log" && /^CommitterName:/   { CommitterName   = substr($0, 2+match($0, ":")) }
script == "log" && /^CommitterEmail:/  { CommitterEmail  = substr($0, 2+match($0, ":")) }
script == "log" && /^CommitterDate:/   { CommitterDate   = substr($0, 2+match($0, ":")) }

### Process output of "git status"
### Changed index?
script == "status" && /^[mMADRCU] /         { modified = 1 }
### Unstaged modifications?
script == "status" && /^ [mMADRCU]/         { modified = 1 }
### Unresolved merge conflicts?
script == "status" && /^[mMADRCU][mMADRCU]/ { modified = 1 }

END {
	### Process output of "git log"
	if (script == "log") {
		### Format dates
		LongDate = substr(CommitterDate, 1, 25)
		DateRAW = substr(LongDate, 1, 10)
		DateISO = DateRAW
		DateTEX = DateISO
		gsub("-", "/", DateTEX)
		Time = substr(LongDate, 12, 14)

		print "%%% This file has been generated by the vc bundle for TeX."
		print "%%% Do not edit this file!"
		print "%%%"

		print "%%% Define Git specific macros."
		print "\\gdef\\GITHash{" Hash "}%"
		print "\\gdef\\GITAbrHash{" AbrHash "}%"
		print "\\gdef\\GITParentHashes{" ParentHashes "}%"
		print "\\gdef\\GITAbrParentHashes{" AbrParentHashes "}%"
		print "\\gdef\\GITAuthorName{" AuthorName "}%"
		print "\\gdef\\GITAuthorEmail{" AuthorEmail "}%"
		print "\\gdef\\GITAuthorDate{" AuthorDate "}%"
		print "\\gdef\\GITCommitterName{" CommitterName "}%"
		print "\\gdef\\GITCommitterEmail{" CommitterEmail "}%"
		print "\\gdef\\GITCommitterDate{" CommitterDate "}%"

		print "%%% Define generic version control macros."
		print "\\gdef\\VCRevision{\\GITAbrHash}%"
		print "\\gdef\\VCAuthor{\\GITAuthorName}%"
		print "\\gdef\\VCDateRAW{" DateRAW "}%"
		print "\\gdef\\VCDateISO{" DateISO "}%"
		print "\\gdef\\VCDateTEX{" DateTEX "}%"
		print "\\gdef\\VCTime{" Time "}%"
		print "\\gdef\\VCModifiedText{\\textcolor{red}{with local modifications!}}%"

		print "%%% Assume clean working copy."
		print "\\gdef\\VCModified{0}%"
		print "\\gdef\\VCRevisionMod{\\VCRevision}%"
	}

	### Process output of "git status"
	if (script=="status") {
		print "%%% Is working copy modified?"
		print "\\gdef\\VCModified{" modified "}%"
		if (modified == 0) {
			print "\\gdef\\VCRevisionMod{\\VCRevision}%"
		} else {
			print "\\gdef\\VCRevisionMod{\\VCRevision~\\VCModifiedText}%"
		}
	}
		
}


svn-unix/vc-svn.awk
# This is file 'vc-svn.awk' from the vc bundle for TeX.
# The original file can be found at CTAN:support/vc.
# This file is Public Domain.
BEGIN {
    
### Process output of "svn info".
		if (script=="info") {
				MaxRevision = 0
		}
		
### Process output of "svn status".
		if (script=="status") {
				modified = 0
		}
		
}



### Process output of "svn info".
### Save all lines of the current record in a corresponding variable.
script=="info" && /^Path:/ { RecPath = $0; next }
script=="info" && /^Name:/ { RecName = $0; next }
script=="info" && /^URL:/ { RecUrl = $0; next }
script=="info" && /^Repository Root:/ { RecRepositoryRoot = $0; next }
script=="info" && /^Repository UUID:/ { RecRepositoryUuid = $0; next }
script=="info" && /^Revision:/ { RecRevision = $2; next }
script=="info" && /^Node Kind:/ { RecNodeKind = $0; next }
script=="info" && /^Last Changed Author:/ { RecLastChangedAuthor = $0; next }
script=="info" && /^Last Changed Rev:/ { RecLastChangedRevision = $0; next }
script=="info" && /^Last Changed Date:/ { RecLastChangedDate = $0; next }
script=="info" && /^[\r]*$/ {
### If revision number of the new record is higher,
### save all lines from the new record.
    if (RecRevision > MaxRevision) {
				MaxPath = RecPath
				MaxName = RecName
				MaxUrl = RecUrl
				MaxRepositoryRoot = RecRepositoryRoot
				MaxRepositoryUuid = RecRepositoryUuid
				MaxRevision = RecRevision
				MaxNodeKind = RecNodeKind
				MaxLastChangedAuthor = RecLastChangedAuthor
				MaxLastChangedRev = RecLastChangedRevision
				MaxLastChangedDate = RecLastChangedDate
    }
}

### Process output of "svn status".
### File with local modifications?
script=="status" && /^[^ ?]/ { modified = 1 }
### File with property modifications?
script=="status" && /^.[^ ]/ { modified = 2 }



END {
		
### Process output of "svn info".
		if (script=="info") {
### Remove possible Windows line endings (e.g., for Msys).
				gsub("\r$", "", MaxPath)
				gsub("\r$", "", MaxName)
				gsub("\r$", "", MaxUrl)
				gsub("\r$", "", MaxRepositoryRoot)
				gsub("\r$", "", MaxRepositoryUuid)
				gsub("\r$", "", MaxRevision)
				gsub("\r$", "", MaxNodeKind)
				gsub("\r$", "", MaxLastChangedAuthor)
				gsub("\r$", "", MaxLastChangedRev)
				gsub("\r$", "", MaxLastChangedDate)
### Escape % characters for TeX compatibility.
				gsub("%", "\\%", MaxUrl)
				gsub("%", "\\%", MaxRepositoryRoot)
### Extract relevant information from variables.
				Path = substr(MaxPath, 2+match(MaxPath, ":"))
				Name = substr(MaxName, 2+match(MaxName, ":"))
				Url = substr(MaxUrl, 2+match(MaxUrl, ":"))
				RepositoryRoot = substr(MaxRepositoryRoot, 2+match(MaxRepositoryRoot, ":"))
				RepositoryUuid = substr(MaxRepositoryUuid, 2+match(MaxRepositoryUuid, ":"))
				Revision = MaxRevision
				NodeKind = substr(MaxNodeKind, 2+match(MaxNodeKind, ":"))
				LastChangedAuthor = substr(MaxLastChangedAuthor, 2+match(MaxLastChangedAuthor, ":"))
				LastChangedRev = substr(MaxLastChangedRev, 2+match(MaxLastChangedRev, ":"))
				LastChangedDate = substr(MaxLastChangedDate, 2+match(MaxLastChangedDate, ":"))
				LongDate = substr(LastChangedDate, 1, 25)
				DateRAW = substr(LongDate, 1, 10)
				DateISO = DateRAW
				DateTEX = DateISO
				gsub("-", "/", DateTEX)
				Time = substr(LongDate, 12, 14)
### Write file identification to vc.tex.
				print "%%% This file has been generated by the vc bundle for TeX."
				print "%%% Do not edit this file!"
				print "%%%"
### Write Subversion specific macros.
				print "%%% Define Subversion specific macros."
				print "\\gdef\\SVNRevision{" Revision "}%"
				print "\\gdef\\SVNLastChangedRev{" LastChangedRev "}%"
				print "\\gdef\\SVNLastChangedAuthor{" LastChangedAuthor "}%"
				print "\\gdef\\SVNLastChangedDate{" LastChangedDate "}%"
				print "\\gdef\\SVNRepositoryUuid{" RepositoryUuid "}%"
				if (full==1) {
						print "\\gdef\\SVNPath{" Path "}%"
						print "\\gdef\\SVNName{" Name "}%"
						print "\\gdef\\SVNUrl{" Url "}%"
						print "\\gdef\\SVNNodeKind{" NodeKind "}%"
						print "\\gdef\\SVNRepositoryRoot{" RepositoryRoot "}%"
				}
### Write generic version control macros.
				print "%%% Define generic version control macros."
				print "\\gdef\\VCRevision{\\SVNRevision}%"
				print "\\gdef\\VCAuthor{\\SVNLastChangedAuthor}%"
				print "\\gdef\\VCDateRAW{" DateRAW "}%"
				print "\\gdef\\VCDateISO{" DateISO "}%"
				print "\\gdef\\VCDateTEX{" DateTEX "}%"
				print "\\gdef\\VCTime{" Time "}%"
				print "\\gdef\\VCModifiedText{\\textcolor{red}{with local modifications!}}%"
				print "%%% Assume clean working copy."
				print "\\gdef\\VCModified{0}%"
				print "\\gdef\\VCRevisionMod{\\VCRevision}%"
				MaxRevision = 0
		}
		
### Process output of "svn status".
		if (script=="status") {
				print "%%% Is working copy modified?"
				print "\\gdef\\VCModified{" modified "}%"
				if (modified==0) {
						print "\\gdef\\VCRevisionMod{\\VCRevision}%"
				} else {
						print "\\gdef\\VCRevisionMod{\\VCRevision~\\VCModifiedText}%"
				}
		}
		
}
